### ISSUE TITLE
Support multiple servers for the same dashboard

### ISSUE THREAD

**Issue by grobie** [prometheus/promdash#180]
As a developer,
Given I deploy my services in multiple zones,
I want to have one dashboard
And switch between the data collected by different prometheus servers in each zone.

**Comment by mwitkow-io** [comment]
A great way to solve it would be through adding template variables in server URLs. It's becoming a real maintenance burden not having this functionality.

**Comment by juliusv** [comment]
@mwitkow-io Agreed on being a maintenance burden (for us too!), and on doing it via template variables. Now we just have to figure out a good UI and JSON encoding for doing that. Currently servers for each expression are saved by their ID. Now they'd need to (optionally) be referred to by a string with some template variables inside instead, which then resolves (hopefully) to an existing server. Biggest challenge is the UI, I think: how to keep the simple case simple with just a dropdown, as it is now, but make it easy to use template variables for servers in individual graphs/expressions instead? I guess which type of UI to use would not be a per-graph/per-expression setting, but a global one (as otherwise it'd become quite cumbersome). So maybe it could be just a global setting which chooses the type of UI to use? Hrm... better ideas?

**Comment by matthiasr** [comment]
Maybe add the templating only on the _URL_ side of the server, then automatically import all variables there into the dashboard and from that point on treat them as if they were a dashboard variable?

**Comment by matthiasr** [comment]
i.e. you'd have one server definition for "prometheus-routing", pointing at `http://prometheus-routine-{{zone}}:9090`, then once you use that in a dashboard that dashboard automatically grows a `zone` template variable.

Obviously you could not use different `zone` variables for other things in the dashboard then but you could use the same one. I think that is a good thing.

**Comment by juliusv** [comment]
@matthiasr Oh, that's a great idea! Thank you. I'm sold! :)

And it even keeps PromDash completely agnostic about the concept of a `zone`. It'd just be an arbitrary template variable used in the server name.

**Comment by mwitkow-io** [comment]
Is anyone working on this? @matthiasr

**Comment by stuartnelson3** [comment]
this issue is not currently being worked on

**Comment by mwitkow-io** [comment]
K, I think we'll hack on it in our spare time.

Can you drop some pointers for "hacking noobs" as to where to start? :)

**Comment by stuartnelson3** [comment]
the [URLGenerator](https://github.com/prometheus/promdash/blob/master/app/assets/javascripts/angular/services/url_generator.js) would be a good place to implement the logic that looks for the double curlies `{{}}`. Add an additional parameter that is an object with currently available template variables (i think those will be available already as `$scope.vars`), and then sub those in using the [VariableInterpolator] like we do in [expression.js.erb](https://github.com/prometheus/promdash/blob/master/app/assets/javascripts/angular/directives/expression.js.erb#L66)

you'll have to scan the list of servers in `shared_graph_behavior.js` and check for any interpolation, then automatically add that as a key to `$scope.vars` if it isn't already there.

that should get you going

**Comment by mwitkow-io** [comment]
So I made a first pass at this. As I'm pretty unused to AngularJS or Ruby, this was both fun and confusing.

**Comment by grobie** [comment]
Great work! That solves the first part of the problem I tried to describe above. I files #375 with an idea how to solve the rest.

**Comment by mwitkow-io** [comment]
And there was much rejoicing! Thanks!

**Comment by grobie** [comment]
This issue specifies the acceptance criteria to switch between different instances/profiles of the same dashboard. Let's wait until #375 is merged before we close this issue here.
